staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -601,6 +704,53 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
 	return info.storages[info.storageVersion].CustomResource, nil
 }
 
+// getOrCreateServingInfoForWrite is getOrCreateServingInfoFor, but blocks CR
+// write requests until the storage version update is processed, aborted, or
+// timed-out.
+// If the storage version update is processed, unblock and allow the write request.
+// If the storage version update is aborted, unblock and reject the write request.
+// If the storage version update timed-out, unblock and allow the write request.
+//
+// If aborted is true, the returned error (403) should be surfaced to the client,
+// indicating that the server forbids the CR request due to the CRD deletion.
+// Otherwise, any non-nil error returned by getOrCreateServingInfoFor may be
+// wrapped by the server as an internal error (500) before surfacing to the client.
+func (r *crdHandler) getOrCreateServingInfoForWrite(uid types.UID, name string) (ret *crdInfo, aborted bool, err error) {
+	ret, err = r.getOrCreateServingInfoFor(uid, name)
+
+	// Surface non-nil error early.
+	if err != nil ||
+		// Return early if the StorageVersionAPI feature gate is disabled.
+		!utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) ||
+		ret.storageVersionUpdate == nil ||
+		ret.storageVersionUpdate.processedCh == nil {
+		return ret, false, err
+	}
+
+	// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+	// to register themselves to the wait group. Ideally the storage version update should
+	// not cause the requests to miss the 1s window; otherwise the requests may
+	// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+	// first CR request establishes the underlying storage).
+	select {
+	case <-ret.storageVersionUpdate.processedCh:
+	case <-ret.storageVersionUpdate.abortedCh:
+		// Storage version update aborted. The caller knows which error
+		// to return to the client (403).
+		return nil, true, nil
+	// Unblock the requests if the storage version update takes a long time, otherwise
+	// CR requests may stack up and overwhelm the API server.
+	// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+	case <-time.After(ret.storageVersionUpdate.timeout.Sub(time.Now())):
```

Comment:
With this timeout we risk an inconsistent storage version state, right? Am still unsure whether is worse: bad storage version object or failing CRD writes.

@liggitt would like to hear your opinion.

Reply:
I'm finding the handling of a CR really hard to reason about in this PR, especially with a long (O(seconds)) delay between constructing the serving info at the top of this method and returning to unblock serving

Reply:
> bad storage version object or failing CRD writes

@sttts I considered failing CRD writes a regression, but I'm okay with that direction if we think it's necessary. Let me add some change in that direction

Reply:
If we cannot write the StorageVersion object, we have big problems anyway. Do we care about CRD writes in this case?